pkg/tbtc/dkg.go

```diff
@@ -224,13 +231,15 @@ func (de *dkgExecutor) setupBroadcastChannel(
 
 // generateSigningGroup executes off-chain protocol for each member controlled
 // by the current operator and upon successful execution of the protocol
-// publishes the result to the chain.
+// publishes the result to the chain. The execution can be delayed by an
+// arbitrary number of blocks using the delayBlocks argument.
 func (de *dkgExecutor) generateSigningGroup(
 	dkgLogger *zap.SugaredLogger,
 	seed *big.Int,
 	memberIndexes []uint8,
 	groupSelectionResult *GroupSelectionResult,
 	startBlock uint64,
+	delayBlocks uint64,
```

Comment:
Does it make sense to pass `startBlock` and `delayBlocks` separately? 

The only reason for that is to log `zap.Uint64("attemptStartBlock", attempt.startBlock)` but I think we could treat the `attemptStartBlock` as the event block with the confirmation delay added. In fact, this is when the attempt starts.

It would also allow us to avoid this indirection: In `pkg/tbtc/tbtc.go` we react on the event and wait for `dkgStartedConfirmationBlocks`. If everything checks, we dispatch the action to `node.go` making the `dkgStartedConfirmationBlocks` an abstract `delayBlocks`. But then, we go back to `pkg/tbtc` and we no longer use `dkgStartedConfirmationBlocks` but an abstract `delayBlocks`. It is like jumping out of the context for a while just to lose some details.

Instead, we could use `confirmationBlock` in `pkg/tbtc/tbtc.go` when calling `node.joinDKGIfEligible` as a `startBlock` (instead of `lastEvent.BlockNumber`).

Reply:
Can you clarify this part:
```
But then, we go back to pkg/tbtc and we no longer use dkgStartedConfirmationBlocks but an abstract delayBlocks. It is like jumping out of the context for a while just to lose some details.
```

Also, the whole story is more complicated. Actually, we need to do two things at the same time:
1. Keep the original start block which must be the same as stored on the contract side. This is a must to produce valid signatures.
2. Delay the off-chain protocol start by the confirmation period

A single start block argument cannot satisfy both requirements. Using `confirmationBlock` based on a uncertain event (the one we confirm) cannot do the trick as the real start block held by the contract may be different. We need to base on the last event we fetch after the confirmation period. In the same time, we need to delay the off-chain protocol execution by the confirmation period length to not fail on the first attempt.

Reply:
I see it now! I missed the usage of `startBlock` in `generateSigningGroup` when calling `de.publishDkgResult`. 

It all makes sense. Can we add this last sentence to `generateSigningGroup` documentation to make it clear:
```
// (...)
// confirming blocks - before executing the off-chain action. Note that the
// startBlock represents the block at which DKG started on-chain. This is
// important for the result submission.
```

Reply:
See [890f7e4](https://github.com/keep-network/keep-core/pull/3468/commits/890f7e4c1871792d37ef3115bd6558cdcf15bbe7)